{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "07a0252d",
   "metadata": {},
   "source": [
    "#### Q1. Explain why we have to use the Exception class while creating a Custom Exception.\n",
    "##### Note: Here Exception class refers to the base class for all the exceptions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d3e2414",
   "metadata": {},
   "source": [
    " - In object-oriented programming, exceptions are used to handle exceptional or unexpected situations that may occur during the execution of a program..\n",
    " \n",
    " - Exception class as the base for creating custom exceptions provides a structured, consistent, and catchable approach to dealing with exceptional situations in your code. It aligns with established programming practices, ensures compatibility, and enhances the clarity and maintainability of your codebase"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b268d1a",
   "metadata": {},
   "source": [
    "## Q2. Write a python program to print Python Exception Hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "78459552",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'Exception'>\n",
      "  <class 'ArithmeticError'>\n",
      "    <class 'FloatingPointError'>\n",
      "    <class 'OverflowError'>\n",
      "    <class 'ZeroDivisionError'>\n",
      "      <class 'decimal.DivisionByZero'>\n",
      "      <class 'decimal.DivisionUndefined'>\n",
      "    <class 'decimal.DecimalException'>\n",
      "      <class 'decimal.Clamped'>\n",
      "      <class 'decimal.Rounded'>\n",
      "        <class 'decimal.Underflow'>\n",
      "        <class 'decimal.Overflow'>\n",
      "      <class 'decimal.Inexact'>\n",
      "        <class 'decimal.Underflow'>\n",
      "        <class 'decimal.Overflow'>\n",
      "      <class 'decimal.Subnormal'>\n",
      "        <class 'decimal.Underflow'>\n",
      "      <class 'decimal.DivisionByZero'>\n",
      "      <class 'decimal.FloatOperation'>\n",
      "      <class 'decimal.InvalidOperation'>\n",
      "        <class 'decimal.ConversionSyntax'>\n",
      "        <class 'decimal.DivisionImpossible'>\n",
      "        <class 'decimal.DivisionUndefined'>\n",
      "        <class 'decimal.InvalidContext'>\n",
      "  <class 'AssertionError'>\n",
      "  <class 'AttributeError'>\n",
      "    <class 'dataclasses.FrozenInstanceError'>\n",
      "  <class 'BufferError'>\n",
      "  <class 'EOFError'>\n",
      "    <class 'asyncio.exceptions.IncompleteReadError'>\n",
      "  <class 'ImportError'>\n",
      "    <class 'ModuleNotFoundError'>\n",
      "      <class 'importlib.metadata.PackageNotFoundError'>\n",
      "    <class 'zipimport.ZipImportError'>\n",
      "  <class 'LookupError'>\n",
      "    <class 'IndexError'>\n",
      "    <class 'KeyError'>\n",
      "      <class 'jupyter_client.kernelspec.NoSuchKernel'>\n",
      "      <class 'IPython.terminal.pt_inputhooks.UnknownBackend'>\n",
      "    <class 'encodings.CodecRegistryError'>\n",
      "  <class 'MemoryError'>\n",
      "  <class 'NameError'>\n",
      "    <class 'UnboundLocalError'>\n",
      "  <class 'OSError'>\n",
      "    <class 'BlockingIOError'>\n",
      "    <class 'ChildProcessError'>\n",
      "    <class 'ConnectionError'>\n",
      "      <class 'BrokenPipeError'>\n",
      "      <class 'ConnectionAbortedError'>\n",
      "      <class 'ConnectionRefusedError'>\n",
      "      <class 'ConnectionResetError'>\n",
      "        <class 'http.client.RemoteDisconnected'>\n",
      "    <class 'FileExistsError'>\n",
      "    <class 'FileNotFoundError'>\n",
      "    <class 'InterruptedError'>\n",
      "      <class 'zmq.error.InterruptedSystemCall'>\n",
      "    <class 'IsADirectoryError'>\n",
      "    <class 'NotADirectoryError'>\n",
      "    <class 'PermissionError'>\n",
      "    <class 'ProcessLookupError'>\n",
      "    <class 'TimeoutError'>\n",
      "    <class 'io.UnsupportedOperation'>\n",
      "    <class 'socket.herror'>\n",
      "    <class 'socket.gaierror'>\n",
      "    <class 'ssl.SSLError'>\n",
      "      <class 'ssl.SSLCertVerificationError'>\n",
      "      <class 'ssl.SSLZeroReturnError'>\n",
      "      <class 'ssl.SSLWantWriteError'>\n",
      "      <class 'ssl.SSLWantReadError'>\n",
      "      <class 'ssl.SSLSyscallError'>\n",
      "      <class 'ssl.SSLEOFError'>\n",
      "    <class 'shutil.Error'>\n",
      "      <class 'shutil.SameFileError'>\n",
      "    <class 'shutil.SpecialFileError'>\n",
      "    <class 'shutil.ExecError'>\n",
      "    <class 'shutil.ReadError'>\n",
      "    <class 'urllib.error.URLError'>\n",
      "      <class 'urllib.error.HTTPError'>\n",
      "      <class 'urllib.error.ContentTooShortError'>\n",
      "    <class 'gzip.BadGzipFile'>\n",
      "  <class 'ReferenceError'>\n",
      "  <class 'RuntimeError'>\n",
      "    <class 'NotImplementedError'>\n",
      "      <class 'zmq.error.ZMQVersionError'>\n",
      "      <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "    <class 'RecursionError'>\n",
      "    <class '_frozen_importlib._DeadlockError'>\n",
      "    <class 'threading.BrokenBarrierError'>\n",
      "    <class 'concurrent.futures._base.BrokenExecutor'>\n",
      "      <class 'concurrent.futures.thread.BrokenThreadPool'>\n",
      "    <class 'asyncio.exceptions.SendfileNotAvailableError'>\n",
      "    <class 'asyncio.exceptions.BrokenBarrierError'>\n",
      "    <class 'pkg_resources.ExtractionError'>\n",
      "    <class '_pydevd_bundle.pydevd_vars.VariableError'>\n",
      "  <class 'StopAsyncIteration'>\n",
      "  <class 'StopIteration'>\n",
      "  <class 'SyntaxError'>\n",
      "    <class 'IndentationError'>\n",
      "      <class 'TabError'>\n",
      "  <class 'SystemError'>\n",
      "    <class 'encodings.CodecRegistryError'>\n",
      "  <class 'TypeError'>\n",
      "    <class 'decimal.FloatOperation'>\n",
      "    <class 'email.errors.MultipartConversionError'>\n",
      "  <class 'ValueError'>\n",
      "    <class 'UnicodeError'>\n",
      "      <class 'UnicodeDecodeError'>\n",
      "      <class 'UnicodeEncodeError'>\n",
      "      <class 'UnicodeTranslateError'>\n",
      "    <class 'io.UnsupportedOperation'>\n",
      "    <class 'json.decoder.JSONDecodeError'>\n",
      "    <class 'ssl.SSLCertVerificationError'>\n",
      "    <class 'binascii.Error'>\n",
      "    <class '_hashlib.UnsupportedDigestmodError'>\n",
      "    <class 'calendar.IllegalMonthError'>\n",
      "    <class 'calendar.IllegalWeekdayError'>\n",
      "    <class 'dateutil.parser._parser.ParserError'>\n",
      "    <class 'email.errors.MessageDefect'>\n",
      "      <class 'email.errors.NoBoundaryInMultipartDefect'>\n",
      "      <class 'email.errors.StartBoundaryNotFoundDefect'>\n",
      "      <class 'email.errors.CloseBoundaryNotFoundDefect'>\n",
      "      <class 'email.errors.FirstHeaderLineIsContinuationDefect'>\n",
      "      <class 'email.errors.MisplacedEnvelopeHeaderDefect'>\n",
      "      <class 'email.errors.MissingHeaderBodySeparatorDefect'>\n",
      "      <class 'email.errors.MultipartInvariantViolationDefect'>\n",
      "      <class 'email.errors.InvalidMultipartContentTransferEncodingDefect'>\n",
      "      <class 'email.errors.UndecodableBytesDefect'>\n",
      "      <class 'email.errors.InvalidBase64PaddingDefect'>\n",
      "      <class 'email.errors.InvalidBase64CharactersDefect'>\n",
      "      <class 'email.errors.InvalidBase64LengthDefect'>\n",
      "      <class 'email.errors.HeaderDefect'>\n",
      "        <class 'email.errors.InvalidHeaderDefect'>\n",
      "        <class 'email.errors.HeaderMissingRequiredValue'>\n",
      "        <class 'email.errors.NonPrintableDefect'>\n",
      "        <class 'email.errors.ObsoleteHeaderDefect'>\n",
      "        <class 'email.errors.NonASCIILocalPartDefect'>\n",
      "        <class 'email.errors.InvalidDateDefect'>\n",
      "    <class 'pygments.util.ClassNotFound'>\n",
      "    <class 'IPython.lib.clipboard.ClipboardEmpty'>\n",
      "    <class 'IPython.core.magics.code.MacroToEdit'>\n",
      "    <class 'plistlib.InvalidFileException'>\n",
      "    <class 'pkg_resources._vendor.more_itertools.more.UnequalIterablesError'>\n",
      "    <class 'pkg_resources.extern.packaging.version.InvalidVersion'>\n",
      "    <class 'pkg_resources._vendor.packaging._manylinux._ELFFileHeader._InvalidELFFileHeader'>\n",
      "    <class 'pkg_resources.extern.packaging.utils.InvalidWheelFilename'>\n",
      "    <class 'pkg_resources.extern.packaging.utils.InvalidSdistFilename'>\n",
      "    <class 'pkg_resources.extern.packaging.specifiers.InvalidSpecifier'>\n",
      "    <class 'pkg_resources.extern.packaging.markers.InvalidMarker'>\n",
      "    <class 'pkg_resources.extern.packaging.markers.UndefinedComparison'>\n",
      "    <class 'pkg_resources.extern.packaging.markers.UndefinedEnvironmentName'>\n",
      "    <class 'pkg_resources.extern.packaging.requirements.InvalidRequirement'>\n",
      "      <class 'pkg_resources.RequirementParseError'>\n",
      "    <class 'packaging.version.InvalidVersion'>\n",
      "  <class 'Warning'>\n",
      "    <class 'BytesWarning'>\n",
      "    <class 'DeprecationWarning'>\n",
      "      <class 'IPython.core.interactiveshell.ProvisionalWarning'>\n",
      "    <class 'EncodingWarning'>\n",
      "    <class 'FutureWarning'>\n",
      "      <class 'IPython.core.completer.ProvisionalCompleterWarning'>\n",
      "    <class 'ImportWarning'>\n",
      "    <class 'PendingDeprecationWarning'>\n",
      "    <class 'ResourceWarning'>\n",
      "    <class 'RuntimeWarning'>\n",
      "      <class 'zmq.asyncio.ProactorSelectorThreadWarning'>\n",
      "      <class 'dateutil.parser._parser.UnknownTimezoneWarning'>\n",
      "      <class 'pkg_resources.PEP440Warning'>\n",
      "    <class 'SyntaxWarning'>\n",
      "    <class 'UnicodeWarning'>\n",
      "    <class 'UserWarning'>\n",
      "      <class 'getpass.GetPassWarning'>\n",
      "      <class 'IPython.core.formatters.FormatterWarning'>\n",
      "    <class 'dateutil.tz.DeprecatedTzFormatWarning'>\n",
      "    <class 'pkg_resources.PkgResourcesDeprecationWarning'>\n",
      "  <class 'ExceptionGroup'>\n",
      "  <class 'warnings._OptionError'>\n",
      "  <class 'runpy._Error'>\n",
      "  <class 're.error'>\n",
      "  <class 'locale.Error'>\n",
      "  <class 'subprocess.SubprocessError'>\n",
      "    <class 'subprocess.CalledProcessError'>\n",
      "    <class 'subprocess.TimeoutExpired'>\n",
      "  <class 'zmq.error.ZMQBaseError'>\n",
      "    <class 'zmq.error.ZMQError'>\n",
      "      <class 'zmq.error.ContextTerminated'>\n",
      "      <class 'zmq.error.Again'>\n",
      "      <class 'zmq.error.InterruptedSystemCall'>\n",
      "    <class 'zmq.error.ZMQBindError'>\n",
      "    <class 'zmq.error.NotDone'>\n",
      "  <class 'copy.Error'>\n",
      "  <class 'struct.error'>\n",
      "  <class '_pickle.PickleError'>\n",
      "    <class '_pickle.PicklingError'>\n",
      "    <class '_pickle.UnpicklingError'>\n",
      "  <class 'pickle._Stop'>\n",
      "  <class 'tokenize.TokenError'>\n",
      "  <class 'tokenize.StopTokenizing'>\n",
      "  <class 'concurrent.futures._base.Error'>\n",
      "    <class 'concurrent.futures._base.CancelledError'>\n",
      "    <class 'concurrent.futures._base.InvalidStateError'>\n",
      "  <class 'socket._GiveupOnSendfile'>\n",
      "  <class 'binascii.Incomplete'>\n",
      "  <class 'inspect.ClassFoundException'>\n",
      "  <class 'inspect.EndOfBlock'>\n",
      "  <class 'asyncio.exceptions.InvalidStateError'>\n",
      "  <class 'asyncio.exceptions.LimitOverrunError'>\n",
      "  <class 'asyncio.queues.QueueEmpty'>\n",
      "  <class 'asyncio.queues.QueueFull'>\n",
      "  <class 'zlib.error'>\n",
      "  <class '_lzma.LZMAError'>\n",
      "  <class 'shutil.RegistryError'>\n",
      "  <class 'shutil._GiveupOnFastCopy'>\n",
      "  <class 'traitlets.traitlets.TraitError'>\n",
      "  <class '_queue.Empty'>\n",
      "  <class 'queue.Full'>\n",
      "  <class 'ctypes.ArgumentError'>\n",
      "  <class '_ctypes.COMError'>\n",
      "  <class 'tornado.concurrent.ReturnValueIgnoredError'>\n",
      "  <class 'argparse.ArgumentError'>\n",
      "  <class 'argparse.ArgumentTypeError'>\n",
      "  <class 'traitlets.config.loader.ConfigError'>\n",
      "    <class 'traitlets.config.loader.ConfigLoaderError'>\n",
      "      <class 'traitlets.config.loader.ArgumentError'>\n",
      "    <class 'traitlets.config.loader.ConfigFileNotFound'>\n",
      "  <class 'traitlets.config.configurable.ConfigurableError'>\n",
      "    <class 'traitlets.config.configurable.MultipleInstanceError'>\n",
      "  <class 'traitlets.config.application.ApplicationError'>\n",
      "  <class 'jupyter_client.channels.InvalidPortNumber'>\n",
      "  <class 'jupyter_client.localinterfaces.NoIPAddresses'>\n",
      "  <class '_csv.Error'>\n",
      "  <class 'zipfile.BadZipFile'>\n",
      "  <class 'zipfile.LargeZipFile'>\n",
      "  <class 'email.errors.MessageError'>\n",
      "    <class 'email.errors.MessageParseError'>\n",
      "      <class 'email.errors.HeaderParseError'>\n",
      "      <class 'email.errors.BoundaryError'>\n",
      "    <class 'email.errors.MultipartConversionError'>\n",
      "    <class 'email.errors.CharsetError'>\n",
      "  <class 'jupyter_client.multikernelmanager.DuplicateKernelError'>\n",
      "  <class 'pydoc.ErrorDuringImport'>\n",
      "  <class 'executing.executing.NotOneValueFound'>\n",
      "  <class 'executing._exceptions.KnownIssue'>\n",
      "  <class 'executing._exceptions.VerifierFailure'>\n",
      "  <class 'pure_eval.utils.CannotEval'>\n",
      "  <class 'pygments.util.OptionError'>\n",
      "  <class 'bdb.BdbQuit'>\n",
      "  <class 'pdb.Restart'>\n",
      "  <class 'IPython.utils.process.FindCmdError'>\n",
      "  <class 'IPython.utils.path.HomeDirError'>\n",
      "  <class 'IPython.core.profiledir.ProfileDirError'>\n",
      "  <class 'IPython.core.error.IPythonCoreError'>\n",
      "    <class 'IPython.core.error.TryNext'>\n",
      "    <class 'IPython.core.error.UsageError'>\n",
      "    <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "  <class 'IPython.core.error.InputRejected'>\n",
      "  <class 'getopt.GetoptError'>\n",
      "  <class 'pygments.filters.ErrorToken'>\n",
      "  <class 'IPython.core.prefilter.PrefilterError'>\n",
      "  <class 'IPython.core.alias.AliasError'>\n",
      "    <class 'IPython.core.alias.InvalidAliasError'>\n",
      "  <class 'sqlite3.Error'>\n",
      "    <class 'sqlite3.InterfaceError'>\n",
      "    <class 'sqlite3.DatabaseError'>\n",
      "      <class 'sqlite3.InternalError'>\n",
      "      <class 'sqlite3.OperationalError'>\n",
      "      <class 'sqlite3.ProgrammingError'>\n",
      "      <class 'sqlite3.IntegrityError'>\n",
      "      <class 'sqlite3.DataError'>\n",
      "      <class 'sqlite3.NotSupportedError'>\n",
      "  <class 'sqlite3.Warning'>\n",
      "  <class 'IPython.core.interactiveshell.SpaceInInput'>\n",
      "  <class 'xml.dom.DOMException'>\n",
      "    <class 'xml.dom.IndexSizeErr'>\n",
      "    <class 'xml.dom.DomstringSizeErr'>\n",
      "    <class 'xml.dom.HierarchyRequestErr'>\n",
      "    <class 'xml.dom.WrongDocumentErr'>\n",
      "    <class 'xml.dom.InvalidCharacterErr'>\n",
      "    <class 'xml.dom.NoDataAllowedErr'>\n",
      "    <class 'xml.dom.NoModificationAllowedErr'>\n",
      "    <class 'xml.dom.NotFoundErr'>\n",
      "    <class 'xml.dom.NotSupportedErr'>\n",
      "    <class 'xml.dom.InuseAttributeErr'>\n",
      "    <class 'xml.dom.InvalidStateErr'>\n",
      "    <class 'xml.dom.SyntaxErr'>\n",
      "    <class 'xml.dom.InvalidModificationErr'>\n",
      "    <class 'xml.dom.NamespaceErr'>\n",
      "    <class 'xml.dom.InvalidAccessErr'>\n",
      "    <class 'xml.dom.ValidationErr'>\n",
      "  <class 'prompt_toolkit.validation.ValidationError'>\n",
      "  <class 'prompt_toolkit.buffer.EditReadOnlyBuffer'>\n",
      "  <class 'prompt_toolkit.buffer._Retry'>\n",
      "  <class 'prompt_toolkit.layout.layout.InvalidLayoutError'>\n",
      "  <class 'prompt_toolkit.renderer.HeightIsUnknownError'>\n",
      "  <class 'IPython.core.guarded_eval.GuardRejection'>\n",
      "  <class 'parso.parser.ParserSyntaxError'>\n",
      "  <class 'parso.parser.InternalParseError'>\n",
      "  <class 'parso.python.diff._PositionUpdatingFinished'>\n",
      "  <class 'jedi.inference.helpers.SimpleGetItemNotFound'>\n",
      "  <class 'jedi.inference.utils.UncaughtAttributeError'>\n",
      "  <class 'jedi.inference.base_value.HasNoContext'>\n",
      "  <class 'jedi.inference.arguments.ParamIssue'>\n",
      "  <class 'jedi.api.exceptions._JediError'>\n",
      "    <class 'jedi.api.exceptions.InternalError'>\n",
      "    <class 'jedi.api.exceptions.WrongVersion'>\n",
      "    <class 'jedi.api.exceptions.RefactoringError'>\n",
      "  <class 'jedi.api.helpers.OnErrorLeaf'>\n",
      "  <class 'jedi.api.environment.InvalidPythonEnvironment'>\n",
      "  <class 'http.client.HTTPException'>\n",
      "    <class 'http.client.NotConnected'>\n",
      "    <class 'http.client.InvalidURL'>\n",
      "    <class 'http.client.UnknownProtocol'>\n",
      "    <class 'http.client.UnknownTransferEncoding'>\n",
      "    <class 'http.client.UnimplementedFileMode'>\n",
      "    <class 'http.client.IncompleteRead'>\n",
      "    <class 'http.client.ImproperConnectionState'>\n",
      "      <class 'http.client.CannotSendRequest'>\n",
      "      <class 'http.client.CannotSendHeader'>\n",
      "      <class 'http.client.ResponseNotReady'>\n",
      "    <class 'http.client.BadStatusLine'>\n",
      "      <class 'http.client.RemoteDisconnected'>\n",
      "    <class 'http.client.LineTooLong'>\n",
      "  <class 'IPython.core.magics.code.InteractivelyDefined'>\n",
      "  <class 'IPython.terminal.embed.KillEmbedded'>\n",
      "  <class 'psutil.Error'>\n",
      "    <class 'psutil.NoSuchProcess'>\n",
      "      <class 'psutil.ZombieProcess'>\n",
      "    <class 'psutil.AccessDenied'>\n",
      "    <class 'psutil.TimeoutExpired'>\n",
      "  <class '_psutil_windows.Error'>\n",
      "  <class '_psutil_windows.TimeoutExpired'>\n",
      "  <class '_psutil_windows.TimeoutAbandoned'>\n",
      "  <class 'tornado.gen.KeyReuseError'>\n",
      "  <class 'tornado.gen.UnknownKeyError'>\n",
      "  <class 'tornado.gen.LeakedCallbackError'>\n",
      "  <class 'tornado.gen.BadYieldError'>\n",
      "  <class 'tornado.gen.ReturnValueIgnoredError'>\n",
      "  <class 'tornado.gen.Return'>\n",
      "  <class 'tornado.queues.QueueEmpty'>\n",
      "  <class 'tornado.queues.QueueFull'>\n",
      "  <class '_pydev_bundle._pydev_saved_modules.DebuggerInitializationError'>\n",
      "  <class 'xml.parsers.expat.ExpatError'>\n",
      "  <class 'xmlrpc.client.Error'>\n",
      "    <class 'xmlrpc.client.ProtocolError'>\n",
      "    <class 'xmlrpc.client.ResponseError'>\n",
      "    <class 'xmlrpc.client.Fault'>\n",
      "  <class 'pywintypes.error'>\n",
      "  <class 'pywintypes.com_error'>\n",
      "  <class 'pythoncom.internal_error'>\n",
      "  <class 'pkg_resources._vendor.pyparsing.exceptions.ParseBaseException'>\n",
      "    <class 'pkg_resources._vendor.pyparsing.exceptions.ParseException'>\n",
      "    <class 'pkg_resources._vendor.pyparsing.exceptions.ParseFatalException'>\n",
      "      <class 'pkg_resources._vendor.pyparsing.exceptions.ParseSyntaxException'>\n",
      "  <class 'pkg_resources._vendor.pyparsing.exceptions.RecursiveGrammarException'>\n",
      "  <class 'pkg_resources.ResolutionError'>\n",
      "    <class 'pkg_resources.VersionConflict'>\n",
      "      <class 'pkg_resources.ContextualVersionConflict'>\n",
      "    <class 'pkg_resources.DistributionNotFound'>\n",
      "    <class 'pkg_resources.UnknownExtra'>\n",
      "  <class '_pydevd_bundle.pydevd_runpy._Error'>\n",
      "  <class '_pydevd_bundle.pydevd_resolver.UnableToResolveVariableException'>\n",
      "  <class '_pydev_bundle.pydev_monkey.InvalidTypeInArgsException'>\n"
     ]
    }
   ],
   "source": [
    "def print_exception_hierarchy(exception_class, indent = 0):\n",
    "    print(\" \" * indent + str(exception_class))\n",
    "    for subclass in exception_class.__subclasses__():\n",
    "        print_exception_hierarchy(subclass, indent + 2)\n",
    "        \n",
    "print_exception_hierarchy(Exception)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "460acf46",
   "metadata": {},
   "source": [
    "### Q3. What errors are defined in the ArithmeticError class? Explain any two with an example.\n",
    "\n",
    "\n",
    "- Ans : In ArithmeticError there are two types of error defined--\n",
    "\n",
    "I ) ZeroDivisionError - This exception is raised when division or modulo operation is performed with zero as the denominator.\n",
    "\n",
    "II ) OverFlowError - This exception occurs when a mathematical operation produces a result that is too large to be represented by the numeric type being used. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d8bcffeb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: division by zero\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    result = 10 / 0  # Division by zero\n",
    "except ZeroDivisionError as e:\n",
    "    print(\"Error:\", e)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7ceedc6",
   "metadata": {},
   "source": [
    "## Q4. Why LookupError class is used? Explain with an example KeyError and IndexError."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a232801a",
   "metadata": {},
   "source": [
    "- LookupError is used as a base class for exceptions related to failed lookups in mappings and sequences, but not for KeyError and IndexError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a17d4bc7",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'keyError' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[4], line 5\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m----> 5\u001b[0m     value \u001b[38;5;241m=\u001b[39m \u001b[43mmy_dict\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mx\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m \u001b[38;5;66;03m#Accesing a non existing key\u001b[39;00m\n\u001b[0;32m      6\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m keyError \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "\u001b[1;31mKeyError\u001b[0m: 'x'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[4], line 6\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m      5\u001b[0m     value \u001b[38;5;241m=\u001b[39m my_dict[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mx\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;66;03m#Accesing a non existing key\u001b[39;00m\n\u001b[1;32m----> 6\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[43mkeyError\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m      7\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mError:\u001b[39m\u001b[38;5;124m\"\u001b[39m, e)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'keyError' is not defined"
     ]
    }
   ],
   "source": [
    "## Key Error\n",
    "\n",
    "my_dict = {'a': 1,'b':2,'c':3}\n",
    "try:\n",
    "    value = my_dict['x'] #Accesing a non existing key\n",
    "except keyError as e:\n",
    "    print(\"Error:\", e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "befb46d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: list index out of range\n"
     ]
    }
   ],
   "source": [
    "# IndexError\n",
    "\n",
    "my_list = [10, 20, 30]\n",
    "try:\n",
    "    element = my_list[5]\n",
    "except IndexError as e:\n",
    "    print(\"Error:\", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc6917e2",
   "metadata": {},
   "source": [
    "## Q5. Explain ImportError. What is ModuleNotFoundError?\n",
    "\n",
    "\n",
    "\n",
    "- Ans --> This exception is more precise than a general ImportError and provides additional information about which module was not found."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b325f01c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ModuleNotFoundError: No module named 'non_existent_module'\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    import non_existent_module\n",
    "    \n",
    "except ModuleNotFoundError as e:\n",
    "    print(\"ModuleNotFoundError:\", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14a0e2d0",
   "metadata": {},
   "source": [
    "# Q6. List down some best practices for exception handling in python."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6c2024b",
   "metadata": {},
   "source": [
    "- Be specific: Catch only expected exceptions.\n",
    "\n",
    "- Use multiple except blocks: Handle different exception types separately.\n",
    "\n",
    "- Use finally: Ensure cleanup code runs, regardless of exceptions.\n",
    "\n",
    "- Avoid bare except: Specify exception types to catch.\n",
    "\n",
    "- Handle specific exceptions: Address different errors accurately.\n",
    "\n",
    "- Avoid deep nesting: Keep try blocks shallow for clarity.\n",
    "\n",
    "- Use context managers: Employ with statements for resource management.\n",
    "\n",
    "- Provide meaningful messages: Clear error messages aid debugging.\n",
    "\n",
    "- Log exceptions: Use logging to record exception details.\n",
    "\n",
    "- Use exceptions for exceptions: Don't overuse for normal flow."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfd43948",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
